File: daphne-server/scripts/management_link.py
"""Independent management-link health evaluation; never formats private identity."""

SPECS = (
    ("operstate", "1", "text_value", 0), ("carrier", "1", "flag_value", 0),
    ("speed", "Mbit/s", "unsigned_value", 0), ("duplex", "1", "text_value", 0),
    ("mtu", "B", "unsigned_value", 0), ("statistics/rx_bytes", "B", "unsigned_value", 64),
    ("statistics/rx_packets", "packet", "unsigned_value", 64),
    ("statistics/rx_errors", "packet", "unsigned_value", 64),
    ("statistics/rx_dropped", "packet", "unsigned_value", 64),
    ("statistics/tx_bytes", "B", "unsigned_value", 64),
    ("statistics/tx_packets", "packet", "unsigned_value", 64),
    ("statistics/tx_errors", "packet", "unsigned_value", 64),
    ("statistics/tx_dropped", "packet", "unsigned_value", 64),
    ("carrier_changes", "count", "unsigned_value", 32),
)


def check_link(link, h, require_all_good=False):
    """Check typed metrics and reported bracket; output only validated values.

    This validates the wire contract, not independent kernel readback or proof
    that two equal samples exclude a transient link change between them.
    """
    def require(ok):
        if not ok:
            raise RuntimeError("Invalid management-link telemetry; private details suppressed")

    require(link.quality in (h.MEASUREMENT_GOOD, h.MEASUREMENT_UNAVAILABLE, h.MEASUREMENT_ERROR) and bool(link.detail))
    good = link.quality == h.MEASUREMENT_GOOD
    require(len(link.observations) == 14 and {i.metric for i in link.observations} == set(range(1, 15)))
    if good:
        require(link.HasField("interface_index") and 0 < link.interface_index <= 0x7fffffff
                and 0 < link.acquisition_started_monotonic_ns <= link.observed_monotonic_ns
                and bool(link.observed_host_unix_ns))
    else:
        require(not link.HasField("interface_index") and not link.observed_monotonic_ns
                and not link.observed_host_unix_ns and not link.link_state_bracket_verified)
    report, selected = {}, {}
    for item in link.observations:
        path, unit, kind, width = SPECS[item.metric - 1]
        require((item.source, item.unit, item.counter_width_bits) == (path, unit, width) and bool(item.detail))
        require(item.quality in (h.MEASUREMENT_GOOD, h.MEASUREMENT_UNAVAILABLE, h.MEASUREMENT_ERROR))
        selected[item.metric] = item
        available = item.quality == h.MEASUREMENT_GOOD
        value = None
        if available:
            require(good and item.WhichOneof("value") == kind
                    and link.acquisition_started_monotonic_ns <= item.acquisition_started_monotonic_ns
                    <= item.observed_monotonic_ns <= link.observed_monotonic_ns and bool(item.observed_host_unix_ns))
            value = getattr(item, kind)
            if item.metric == h.MANAGEMENT_LINK_OPERSTATE:
                require(value in ("unknown", "notpresent", "down", "lowerlayerdown", "testing", "dormant", "up"))
            elif item.metric == h.MANAGEMENT_LINK_DUPLEX:
                require(value in ("half", "full"))
            elif item.metric == h.MANAGEMENT_LINK_SPEED_MBPS:
                require(0 < value < 0xffffffff)
            elif item.metric == h.MANAGEMENT_LINK_MTU_BYTES:
                require(0 < value <= 0xffffffff)
            elif width:
                require(0 <= value < 1 << width)
        else:
            require(item.WhichOneof("value") is None and not item.observed_monotonic_ns and not item.observed_host_unix_ns)
        report[h.ManagementLinkMetric.Name(item.metric)] = {
            "quality": h.MeasurementQuality.Name(item.quality), "value": value, "unit": unit}
    state, carrier = selected[h.MANAGEMENT_LINK_OPERSTATE], selected[h.MANAGEMENT_LINK_CARRIER]
    if link.link_state_bracket_verified:
        require(state.quality == carrier.quality == h.MEASUREMENT_GOOD)
    negotiation = (link.link_state_bracket_verified and state.text_value == "up" and carrier.flag_value)
    if not negotiation:
        require(all(selected[n].quality != h.MEASUREMENT_GOOD for n in (h.MANAGEMENT_LINK_SPEED_MBPS, h.MANAGEMENT_LINK_DUPLEX)))
    if require_all_good:
        require(good and all(i.quality == h.MEASUREMENT_GOOD for i in link.observations) and negotiation)
    return {"quality": h.MeasurementQuality.Name(link.quality), "metrics": report,
            "observed_monotonic_ns": link.observed_monotonic_ns,
            "link_state_bracket_verified": link.link_state_bracket_verified,
            "counter_epoch": "unobserved", "scope": "management NIC only; not reachability or Hermes delivery"}


def health_state(network, h, now):
    unknown, passed, failed = h.HEALTH_CHECK_UNKNOWN, h.HEALTH_CHECK_PASS, h.HEALTH_CHECK_FAIL

    def fresh(quality, observed):
        return quality == h.MEASUREMENT_GOOD and 0 < observed <= now and now - observed <= 5_000_000_000

    if not fresh(network.quality, network.observed_monotonic_ns) or not network.HasField("present"):
        return unknown
    if (not network.present or (network.HasField("interface_up") and not network.interface_up)
            or (network.HasField("running_flag") and not network.running_flag)):
        return failed
    if not network.HasField("interface_up") or not network.HasField("running_flag"):
        return unknown
    link = network.link
    if (not network.HasField("interface_index") or not link.HasField("interface_index")
            or not network.interface_index or link.interface_index != network.interface_index
            or not link.link_state_bracket_verified or not fresh(link.quality, link.observed_monotonic_ns)
            or not 0 < network.acquisition_started_monotonic_ns <= link.acquisition_started_monotonic_ns
            <= link.observed_monotonic_ns <= network.observed_monotonic_ns):
        return unknown
    selected = {}
    for item in link.observations:
        if item.metric not in (h.MANAGEMENT_LINK_OPERSTATE, h.MANAGEMENT_LINK_CARRIER):
            continue
        if (item.metric in selected or not fresh(item.quality, item.observed_monotonic_ns)
                or not link.acquisition_started_monotonic_ns <= item.acquisition_started_monotonic_ns
                <= item.observed_monotonic_ns <= link.observed_monotonic_ns):
            return unknown
        selected[item.metric] = item
    state, carrier = selected.get(h.MANAGEMENT_LINK_OPERSTATE), selected.get(h.MANAGEMENT_LINK_CARRIER)
    if (state is None or carrier is None or state.WhichOneof("value") != "text_value"
            or carrier.WhichOneof("value") != "flag_value"
            or state.text_value not in ("unknown", "notpresent", "down", "lowerlayerdown", "testing", "dormant", "up")):
        return unknown
    if not carrier.flag_value:
        return failed
    if state.text_value == "unknown":
        return unknown
    return passed if state.text_value == "up" else failed

File: daphne-server/srcs/server_controller/management_link.cpp
#include "server_controller/management_link.hpp"
#include "server_controller/board_monitor.hpp"

#include <array>
#include <cerrno>
#include <limits>
#include <stdexcept>
#include <system_error>
#ifdef __linux__
#include <fcntl.h>
#include <linux/magic.h>
#include <sys/vfs.h>
#include <unistd.h>
#endif

namespace daphne_sc {
namespace {
using Observation = daphne::ManagementLinkObservation;
struct Spec { daphne::ManagementLinkMetric metric; const char* path; const char* unit; uint32_t width; };
constexpr std::array<Spec, 14> specs{{
    {daphne::MANAGEMENT_LINK_OPERSTATE, "operstate", "1", 0},
    {daphne::MANAGEMENT_LINK_CARRIER, "carrier", "1", 0},
    {daphne::MANAGEMENT_LINK_SPEED_MBPS, "speed", "Mbit/s", 0},
    {daphne::MANAGEMENT_LINK_DUPLEX, "duplex", "1", 0},
    {daphne::MANAGEMENT_LINK_MTU_BYTES, "mtu", "B", 0},
    {daphne::MANAGEMENT_LINK_RX_BYTES, "statistics/rx_bytes", "B", 64},
    {daphne::MANAGEMENT_LINK_RX_PACKETS, "statistics/rx_packets", "packet", 64},
    {daphne::MANAGEMENT_LINK_RX_ERRORS, "statistics/rx_errors", "packet", 64},
    {daphne::MANAGEMENT_LINK_RX_DROPPED, "statistics/rx_dropped", "packet", 64},
    {daphne::MANAGEMENT_LINK_TX_BYTES, "statistics/tx_bytes", "B", 64},
    {daphne::MANAGEMENT_LINK_TX_PACKETS, "statistics/tx_packets", "packet", 64},
    {daphne::MANAGEMENT_LINK_TX_ERRORS, "statistics/tx_errors", "packet", 64},
    {daphne::MANAGEMENT_LINK_TX_DROPPED, "statistics/tx_dropped", "packet", 64},
    {daphne::MANAGEMENT_LINK_CARRIER_CHANGES, "carrier_changes", "count", 32}}};
class Unavailable : public std::runtime_error { using std::runtime_error::runtime_error; };
std::string token(std::string input) {
  if (input.empty() || input.size() > 64) throw std::runtime_error("Invalid link input size");
  if (input.back() == '\n') input.pop_back();
  if (input.empty() || input.find_first_not_of("abcdefghijklmnopqrstuvwxyz0123456789-") != std::string::npos)
    throw std::runtime_error("Invalid link token");
  return input;
}
uint64_t integer(const std::string& value) {
  if (value.empty() || value.size() > 20 || value.find_first_not_of("0123456789") != std::string::npos)
    throw std::runtime_error("Invalid link integer");
  return std::stoull(value);
}
void fail(Observation& item, daphne::MeasurementQuality quality, const char* reason) {
  item.clear_value(); item.clear_observed_monotonic_ns(); item.clear_observed_host_unix_ns();
  item.set_quality(quality); item.set_detail(reason);
}
void decode(Observation& item, const std::string& text) {
  switch (item.metric()) {
    case daphne::MANAGEMENT_LINK_OPERSTATE:
      if (text != "unknown" && text != "notpresent" && text != "down" && text != "lowerlayerdown" &&
          text != "testing" && text != "dormant" && text != "up") throw std::runtime_error("Invalid operstate");
      item.set_text_value(text); return;
    case daphne::MANAGEMENT_LINK_CARRIER:
      if (text != "0" && text != "1") throw std::runtime_error("Invalid carrier");
      item.set_flag_value(text == "1"); return;
    case daphne::MANAGEMENT_LINK_DUPLEX:
      if (text == "unknown") throw Unavailable("Unknown duplex");
      if (text != "half" && text != "full") throw std::runtime_error("Invalid duplex");
      item.set_text_value(text); return;
    case daphne::MANAGEMENT_LINK_SPEED_MBPS:
      if (text == "-1" || text == "4294967295") throw Unavailable("Unknown speed");
      break;
    default: break;
  }
  const auto value = integer(text);
  if ((item.metric() == daphne::MANAGEMENT_LINK_SPEED_MBPS || item.metric() == daphne::MANAGEMENT_LINK_MTU_BYTES) &&
      (!value || value > UINT32_MAX)) throw std::runtime_error("Invalid link setting range");
  if (item.counter_width_bits() == 32 && value > UINT32_MAX) throw std::runtime_error("Link counter overflow");
  item.set_unsigned_value(value);
}
bool unavailable_errno(const std::error_code& code) {
  return code == std::errc::no_such_file_or_directory || code == std::errc::no_such_device ||
      code == std::errc::permission_denied || code == std::errc::operation_not_supported ||
      code == std::errc::invalid_argument; // sysfs carrier/speed can reject an admin-down device.
}
Observation collect(ManagementLinkIo& io, const Spec& spec) {
  Observation item;
  item.set_metric(spec.metric); item.set_source(spec.path); item.set_unit(spec.unit);
  item.set_counter_width_bits(spec.width);
  try {
    item.set_acquisition_started_monotonic_ns(io.now().monotonic_ns);
    decode(item, token(io.read_file(spec.path, 64)));
    const auto end = io.now();
    if (!item.acquisition_started_monotonic_ns() || end.monotonic_ns < item.acquisition_started_monotonic_ns())
      throw std::runtime_error("Invalid link acquisition clock");
    item.set_quality(daphne::MEASUREMENT_GOOD); item.set_observed_monotonic_ns(end.monotonic_ns);
    item.set_observed_host_unix_ns(end.host_unix_ns);
    item.set_detail(spec.width ? "Cumulative driver-exported count; reset epoch and hardware counter width not observed" :
                                "Sampled Linux attribute; no end-to-end connectivity inference");
  } catch (const Unavailable&) {
    fail(item, daphne::MEASUREMENT_UNAVAILABLE, "Driver reports unknown link setting");
  } catch (const std::system_error& e) {
    fail(item, unavailable_errno(e.code()) ? daphne::MEASUREMENT_UNAVAILABLE : daphne::MEASUREMENT_ERROR,
         "Link attribute unavailable or read failed; private details suppressed");
  } catch (const std::exception&) {
    fail(item, daphne::MEASUREMENT_ERROR, "Invalid link data or observation clock");
  }
  return item;
}
void invalidate(daphne::ManagementLinkStatus& result, daphne::MeasurementQuality quality) {
  result.set_quality(quality); result.set_detail("Management interface bracket unavailable or changed; no values retained");
  result.clear_interface_index(); result.clear_observed_monotonic_ns(); result.clear_observed_host_unix_ns();
  result.set_link_state_bracket_verified(false);
  for (auto& item : *result.mutable_observations()) fail(item, quality, "Management interface bracket not verified");
}
daphne::ManagementLinkStatus empty_status() {
  daphne::ManagementLinkStatus result;
  for (const auto& spec : specs) {
    auto* item = result.add_observations();
    item->set_metric(spec.metric); item->set_source(spec.path); item->set_unit(spec.unit); item->set_counter_width_bits(spec.width);
  }
  return result;
}
#ifdef __linux__
class LinuxLink final : public ManagementLinkIo {
 public:
  explicit LinuxLink(const std::string& interface) {
    if (interface.empty() || interface.size() >= 16 || interface == "." || interface == ".." ||
        interface.find_first_not_of("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_.:-") != std::string::npos)
      throw std::runtime_error("Invalid selected interface");
    fd_ = open(("/sys/class/net/" + interface).c_str(), O_RDONLY | O_DIRECTORY | O_CLOEXEC);
    if (fd_ < 0) throw std::system_error(errno, std::generic_category());
    struct statfs info{};
    if (fstatfs(fd_, &info) != 0 || info.f_type != SYSFS_MAGIC) {
      close(fd_); fd_ = -1; throw std::runtime_error("Not a sysfs interface directory");
    }
  }
  ~LinuxLink() override { if (fd_ >= 0) close(fd_); }
  std::string read_file(const char* path, size_t maximum) override {
    bool allowed = std::string(path) == "ifindex";
    for (const auto& spec : specs) allowed |= std::string(path) == spec.path;
    if (!allowed || maximum != 64) throw std::runtime_error("Unapproved link attribute");
    const int input = openat(fd_, path, O_RDONLY | O_CLOEXEC | O_NOFOLLOW | O_NONBLOCK);
    if (input < 0) throw std::system_error(errno, std::generic_category());
    std::string data;
    try {
      char buffer[65];
      for (;;) {
        const auto count = read(input, buffer, maximum + 1 - data.size());
        if (count < 0) {
          if (errno == EINTR) continue;
          throw std::system_error(errno, std::generic_category());
        }
        if (!count) break;
        data.append(buffer, static_cast<size_t>(count));
        if (data.size() > maximum) throw std::runtime_error("Oversized link attribute");
      }
    } catch (...) { close(input); throw; }
    close(input); return data;
  }
  ManagementLinkTime now() override { return {host_unix_time_ns(), monotonic_time_ns()}; }
 private:
  int fd_ = -1;
};
#endif
}  // namespace

daphne::ManagementLinkStatus read_management_link(ManagementLinkIo& io, uint32_t expected_index) {
  auto result = empty_status();
  try {
    result.set_acquisition_started_monotonic_ns(io.now().monotonic_ns);
    if (!expected_index || expected_index > INT32_MAX ||
        integer(token(io.read_file("ifindex", 64))) != expected_index) throw std::runtime_error("Wrong interface index");
    result.set_interface_index(expected_index);
    for (size_t n = 0; n < specs.size(); ++n) *result.mutable_observations(n) = collect(io, specs[n]);
    const auto state_after = collect(io, specs[0]), carrier_after = collect(io, specs[1]);
    const auto& state = result.observations(0);
    const auto& carrier = result.observations(1);
    const bool stable = state.quality() == daphne::MEASUREMENT_GOOD && carrier.quality() == daphne::MEASUREMENT_GOOD &&
        state_after.quality() == daphne::MEASUREMENT_GOOD && carrier_after.quality() == daphne::MEASUREMENT_GOOD &&
        state.text_value() == state_after.text_value() && carrier.flag_value() == carrier_after.flag_value();
    result.set_link_state_bracket_verified(stable);
    if (!stable || !carrier.flag_value() || state.text_value() != "up") {
      for (int n : {2, 3}) fail(*result.mutable_observations(n), daphne::MEASUREMENT_UNAVAILABLE,
          "Current negotiation not established by matching carrier-up/operstate-up bracket");
    }
    if (integer(token(io.read_file("ifindex", 64))) != expected_index) throw std::runtime_error("Interface index changed");
    const auto end = io.now();
    if (!result.acquisition_started_monotonic_ns() || end.monotonic_ns < result.acquisition_started_monotonic_ns())
      throw std::runtime_error("Invalid link bracket clock");
    for (const auto& item : result.observations()) if (item.quality() == daphne::MEASUREMENT_GOOD &&
        (item.acquisition_started_monotonic_ns() < result.acquisition_started_monotonic_ns() || item.observed_monotonic_ns() > end.monotonic_ns))
      throw std::runtime_error("Invalid metric bracket clock");
    for (const auto* item : {&state_after, &carrier_after}) if (item->quality() == daphne::MEASUREMENT_GOOD &&
        (item->acquisition_started_monotonic_ns() < result.acquisition_started_monotonic_ns() || item->observed_monotonic_ns() > end.monotonic_ns))
      throw std::runtime_error("Invalid final state bracket clock");
    result.set_quality(daphne::MEASUREMENT_GOOD); result.set_observed_monotonic_ns(end.monotonic_ns);
    result.set_observed_host_unix_ns(end.host_unix_ns);
    result.set_detail("Matching ifindex bracket; inspect individual metric quality. Sequential, not atomic; counter reset epoch unobserved");
  } catch (const std::system_error& e) {
    invalidate(result, unavailable_errno(e.code()) ? daphne::MEASUREMENT_UNAVAILABLE : daphne::MEASUREMENT_ERROR);
  } catch (const std::exception&) { invalidate(result, daphne::MEASUREMENT_ERROR); }
  return result;
}

daphne::ManagementLinkStatus read_management_link_linux(const std::string& interface, uint32_t expected_index) {
  auto result = empty_status();
  try {
#ifdef __linux__
    LinuxLink io(interface);
    return read_management_link(io, expected_index);
#else
    (void)interface; (void)expected_index;
    invalidate(result, daphne::MEASUREMENT_UNAVAILABLE);
#endif
  } catch (const std::system_error& e) {
    invalidate(result, unavailable_errno(e.code()) ? daphne::MEASUREMENT_UNAVAILABLE : daphne::MEASUREMENT_ERROR);
  } catch (const std::exception&) { invalidate(result, daphne::MEASUREMENT_ERROR); }
  return result;
}
}  // namespace daphne_sc
